projects/charger/src/generic_can_network.c

```diff
@@ -0,0 +1,68 @@
+#include "generic_can_network.h"
+
+#include <stddef.h>
+#include <stdint.h>
+#include <string.h>
+
+#include "can.h"
+#include "generic_can.h"
+#include "generic_can_helpers.h"
+#include "generic_can_msg.h"
+#include "log.h"
+#include "status.h"
+
+static GenericCanInterface s_interface;
+
+// CanRxHandlerCb
+static StatusCode prv_generic_can_network_rx_handler(const CANMessage *msg, void *context,
+                                                     CANAckStatus *ack_reply) {
+  (void)ack_reply;
+  GenericCanNetwork *gcn = context;
+  GenericCanMsg generic_msg = { 0 };
+  can_message_to_generic_can_message(msg, &generic_msg);
+  for (size_t i = 0; i < NUM_GENERIC_CAN_RX_HANDLERS; i++) {
+    if (gcn->base.rx_storage[i].rx_handler != NULL &&
+        (generic_msg.id & gcn->base.rx_storage[i].mask) == gcn->base.rx_storage[i].filter) {
+      gcn->base.rx_storage[i].rx_handler(&generic_msg, gcn->base.rx_storage[i].context);
+      break;
+    }
+  }
+  return STATUS_CODE_OK;
+}
+
+// tx
+static StatusCode prv_tx(const GenericCan *can, const GenericCanMsg *msg) {
+  GenericCanNetwork *gcn = (GenericCanNetwork *)can;
+  if (gcn->base.interface != &s_interface) {
+    return status_msg(STATUS_CODE_INVALID_ARGS, "GenericCan not aligned to GenericCanNetwork.");
+  }
+  CANMessage can_msg = { 0 };
+  status_ok_or_return(generic_can_msg_to_can_message(msg, &can_msg));
+
+  return can_transmit(&can_msg, NULL);
+}
+
+// register_rx
+static StatusCode prv_register_rx(GenericCan *can, GenericCanRx rx_handler, uint32_t mask,
+                                  uint32_t filter, bool extended, void *context) {
+  GenericCanNetwork *gcn = (GenericCanNetwork *)can;
+  if (gcn->base.interface != &s_interface) {
+    return status_msg(STATUS_CODE_INVALID_ARGS, "GenericCan not aligned to GenericCanNetwork.");
+  } else if (extended) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+  return generic_can_helpers_register_rx(can, rx_handler, mask, filter, context);
+}
+
+StatusCode generic_can_network_init(GenericCanNetwork *can_network) {
```

Comment:
You should also add a fully permissive filter in case another module has added a non-permissive one.

Reply:
Done.